File: 375__Guess_Number_Higher_or_Lower_II.py
class Solution(object):
    def getMoneyAmount(self, n):
        """
        :type n: int
        :rtype: int
        """
        dp = [[0] * (n + 1) for _ in range(n + 1)]
        for length in range(2, n + 1):
            for start in range(1, n - length + 2):
                end = start + length - 1
                tempmin = float("inf")
                for mid in range(start, end):
                    temp = mid + max(dp[start][mid - 1], dp[mid + 1][end])
                    tempmin = min(tempmin, temp)
                dp[start][end] = tempmin
        return dp[1][-1]

File: Java/375__Guess_Number_Higher_or_Lower_II.java
public class Solution {
    public int getMoneyAmount(int n) {
        int[][] dp = new int[n + 1][n + 1];
        for (int len = 2; len <= n; len++) {
            for (int start = 1; start <= n - len + 1; start++) {
                int min_result = Integer.MAX_VALUE;
                for (int piv = start; piv < start + len - 1; piv++) {
                    int val = piv + Math.max(dp[start][piv - 1], dp[piv + 1][start + len - 1]);
                    min_result = Math.min(val, min_result);
                }
                dp[start][start + len - 1] = min_result;
            }
        }
        return dp[1][n];
    }
}
